### PR TITLE
Add Silex provider

### PR DESCRIPTION
**Pull request by thewilkybarkid**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -16,6 +16,21 @@ Installation
 
 `composer require elife/content-negotiator`
 
+Set up
+------
+
+### Silex
+
+```php
+use eLife\ContentNegotiator\Silex\ContentNegotiationProvider;
```
Comment (giorgiosironi) [inline README.md]: the example could benefit from a `use` for `Accept` to understand where it does come from (`Response` seems obvious but maybe it's not)

File: src/Silex/ContentNegotiationProvider.php [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+        $app['negotiate.accept_language'] = function () {
+            return function (string ...$types) {
+                return function (Request $request, Container $app) use ($types) {
+                    return $app['content_negotiator.accept_language']->negotiate($request, $types);
+                };
+            };
+        };
+
+        if (interface_exists(ArgumentValueResolverInterface::class)) {
```
Comment (giorgiosironi) [inline src/Silex/ContentNegotiationProvider.php]: what's the use case in which this interface doesn't exist? I guess old versions of `symfony/http-kernel` from the tests

File: test/Symfony/TypeAttributeValueResolverTest.php [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    public function it_does_not_support_if_the_type_is_different()
+    {
+        $resolver = new TypeAttributeValueResolver('type', 'attribute');
+
+        $request = new Request();
+        $request->attributes->set('attribute', 'value');
+        $argument = new ArgumentMetadata('name', 'foo', false, false, null);
+
+        $this->assertFalse($resolver->supports($request, $argument));
+    }
```
Comment (giorgiosironi) [inline test/Symfony/TypeAttributeValueResolverTest.php]: `supports()` and `resolve()` seem similar in complexity but only `supports()` seems covered?

File: test/Silex/ContentNegotiationProviderTest.php [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+
+    /**
+     * @test
+     * @dataProvider acceptLanguageProvider
+     */
+    public function it_negotiates_the_accept_language_header(string $header = null, int $statusCode, string $content = null)
+    {
+        $client = $this->createClient();
+
+        $path = interface_exists(ArgumentValueResolverInterface::class) ? '/accept-language' : '/accept-language';
```
Comment (giorgiosironi) [inline test/Silex/ContentNegotiationProviderTest.php]: `$path` is the same no matter the result of the `interface_exists()` call?

File: README.md (hunk 1 above)
Comment (thewilkybarkid) [reply README.md]: Added details.

File: src/Silex/ContentNegotiationProvider.php (hunk 2 above)
Comment (thewilkybarkid) [reply src/Silex/ContentNegotiationProvider.php]: Added comment.

File: test/Symfony/TypeAttributeValueResolverTest.php (hunk 3 above)
Comment (thewilkybarkid) [reply test/Symfony/TypeAttributeValueResolverTest.php]: Added missing assertion.

File: src/Silex/ContentNegotiationProvider.php [hunk 5]
```diff
@@ -61,31 +61,31 @@ public function register(Container $app)
         $app['negotiate.accept'] = function () {
             return function (string ...$types) {
                 return function (Request $request, Container $app) use ($types) {
-                    return $app['content_negotiator.accept']->negotiate($request, $types);
+                    $app['content_negotiator.accept']->negotiate($request, $types);
```
Comment (thewilkybarkid) [inline src/Silex/ContentNegotiationProvider.php]: Maybe `ContentNegotiator` should actually return the match and not set it on a request attribute? That could be done here instead.

File: src/Silex/ContentNegotiationProvider.php [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+            return new CharsetNegotiator();
+        };
+
+        $app['negotiator.accept_encoding'] = function () {
+            return new EncodingNegotiator();
+        };
+
+        $app['negotiator.accept_language'] = function () {
+            return new LanguageNegotiator();
+        };
```
Comment (giorgiosironi) [inline src/Silex/ContentNegotiationProvider.php]: not sure what the double configuration means `negotiator.*` vs `content_negotiator.*`, multiple versions support?

File: src/Silex/ContentNegotiationProvider.php (hunk 5 above)
Comment (giorgiosironi) [reply src/Silex/ContentNegotiationProvider.php]: The current solution seems to minimize the code duplicated between different controllers. This part is not replicated so a change here should be neutral in that respect. If it improved readability here (multiple anonymous functions nested inside each other) it's a net positive.

File: src/Silex/ContentNegotiationProvider.php (hunk 6 above)
Comment (thewilkybarkid) [reply src/Silex/ContentNegotiationProvider.php]: `willdurand/negotiation` services vs wrappers provided by this library. Couldn't think of a better naming strategy, though aware that it's not that clear.

Comment (giorgiosironi) [reply src/Silex/ContentNegotiationProvider.php]: Is it really necessary to expose the `willdurand/negotiation` in `$app`? Can they be hidden from clients instead? Perhaps prefix with `_` their names?

Comment (thewilkybarkid) [reply src/Silex/ContentNegotiationProvider.php]: No, could just remove them as services entirely.

**Pull request merged**